### PR TITLE
Support specifiying "pip" packages in the CLI

### PR DESCRIPTION
**Pull request by chmeliik**
CLOUDBLD-11105

Add everything needed to support pip except for the `fetch_pip_source` function which would actually do the work

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [ ] ~Docs updated (if applicable)~
- [ ] ~Docs links in the code are still valid (if docs were updated)~
- [ ] ~[Issue](https://github.com/containerbuildsystem/cachi2/issues/13) that tracks breaking changes to the OSBS2 integration is updated (if applicable)~

### CODE REVIEW

File: cachi2/core/__init__.py [hunk 1]
```diff
@@ -1,4 +0,0 @@
-# The following imports are functions that serve as the interface
-# to use Cachi2. No module should need to access files from
-# cachi2.core directly.
-from cachi2.core.package_managers.gomod import fetch_gomod_source  # noqa: F401
```
Comment (chmeliik) [inline cachi2/core/__init__.py:4]: Noticed this a bit late :smile: 

Do we still want to try and achieve this?

File: cachi2/core/resolver.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+def resolve_packages(request: Request) -> RequestOutput:
+    """Run all requested package managers, return their combined output."""
+    # de-dupe but preserve order
+    requested_types = dict.fromkeys(pkg.type for pkg in request.packages).keys()
+    unsupported_types = requested_types - _package_managers.keys()
+    if unsupported_types:
+        raise UnsupportedFeature(
+            f"Package manager(s) not yet supported: {', '.join(sorted(unsupported_types))}",
+            # unknown package managers shouldn't get past input validation
+            solution="But the good news is that we're already working on it!",
```
Comment (brunoapimentel) [inline cachi2/core/resolver.py:27]: "But the good news is that YOU can work on it." :laughing: 

On a serious note, an ideia is to point them to the part of the README that shows the plans for package manager support (aka what CachiTO supports).

Comment (chmeliik) [reply cachi2/core/resolver.py:27]: The thing is, the only way you get this error message is if we already have that package manager defined as a valid input (i.e. already working on it)

Comment (chmeliik) [reply cachi2/core/resolver.py:27]: Should I try to allow arbitrary strings as the package manager to make sure it only fails here?

Comment (chmeliik) [reply cachi2/core/resolver.py:27]: No such luck for the Union-based solution. Best I could come up with was this

```diff
diff --git a/cachi2/core/models/input.py b/cachi2/core/models/input.py
index 47d48c1..3300b17 100644
--- a/cachi2/core/models/input.py
+++ b/cachi2/core/models/input.py
@@ -53,7 +53,7 @@ Flag = Literal["cgo-disable", "force-gomod-tidy", "gomod-vendor", "gomod-vendor-
 class _PackageInputBase(pydantic.BaseModel, extra="forbid"):
     """Common input attributes accepted for all package types."""
 
-    type: PackageManagerType
+    type: str
     path: Path = Path(".")
 
     @pydantic.validator("path")
@@ -87,8 +87,14 @@ class PipPackageInput(_PackageInputBase):
         return check_sane_relpath(path)
 
 
+class UnsupportedPackageInput(_PackageInputBase):
+    """TODO."""
+
+    type: Literal["npm", "yarn", "rubygems"]
+
+
 PackageInput = Annotated[
-    Union[GomodPackageInput, PipPackageInput],
+    Union[GomodPackageInput, PipPackageInput, UnsupportedPackageInput],
     # https://pydantic-docs.helpmanual.io/usage/types/#discriminated-unions-aka-tagged-unions
     pydantic.Field(discriminator="type"),
 ]
```

File: cachi2/core/resolver.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+
+supported_package_managers = list(_package_managers)
+
+
+def resolve_packages(request: Request) -> RequestOutput:
+    """Run all requested package managers, return their combined output."""
+    # de-dupe but preserve order
+    requested_types = dict.fromkeys(pkg.type for pkg in request.packages).keys()
```
Comment (brunoapimentel) [inline cachi2/core/resolver.py:21]: I was trying to figure out why we need to dedup and preserve order here, but I couldn't. :sweat_smile: 

Can we use a set here instead?

File: cachi2/core/models/output.py [hunk 4]
```diff
@@ -83,11 +82,18 @@ def unique_packages(cls, packages: list[Package]) -> list[Package]:
...
         """Sort and de-duplicate environment variables by name."""
         return unique_sorted(env_vars, by=lambda env_var: env_var.name)
 
     @classmethod
     def empty(cls):
         """Return an empty RequestOutput."""
         return cls(packages=[], environment_variables=[])
+
+
+def merge_outputs(outputs: Iterable[RequestOutput]) -> RequestOutput:
```
Comment (brunoapimentel) [inline cachi2/core/models/output.py:95]: Does it make sense to move this to the `resolver.py` module?

File: cachi2/core/resolver.py (hunk 3 above)
Comment (chmeliik) [reply cachi2/core/resolver.py:21]: I figured if the user specifies `--package pip --package gomod`, they might expect pip packages to be processed first. But `--package pip --package gomod --package pip` would end up breaking their expectations and going pip -> pip -> gomod anyway, so might as well use a `sorted(set())`

File: cachi2/core/models/output.py (hunk 4 above)
Comment (chmeliik) [reply cachi2/core/models/output.py]: Moved and dropped the unit test, it seemed like overkill for a function this simple anyway

File: cachi2/core/resolver.py (hunk 3 above)
Comment (chmeliik) [reply cachi2/core/resolver.py]: Changed to a set

File: cachi2/core/models/output.py (hunk 4 above)
Comment (brunoapimentel) [reply cachi2/core/models/output.py]: I tend to like small, focused functions, anyway :sweat_smile: 
I'd argue that if they're private, they don't necessarily need to be unit tested separately, though.

**Comment by chmeliik** [comment]
Rebased on main

**Review by lkolacek — APPROVED** [review]
Looks great! :+1: Thank you Adam

**Pull request merged**